hub/util/transform.py

```diff
@@ -0,0 +1,261 @@
+        ds_out[tensor].chunk_engine.cache[meta_key] = tensor_meta
+    ds_out.flush()
+
+
+def combine_metas(ds_tensor_meta: TensorMeta, worker_tensor_meta: TensorMeta) -> None:
+    """Combines the dataset's tensor meta with a single worker's tensor meta."""
+    # if tensor meta is empty, copy attributes from current_meta
+    if len(ds_tensor_meta.max_shape) == 0 or ds_tensor_meta.dtype is None:
+        ds_tensor_meta.dtype = worker_tensor_meta.dtype
+        ds_tensor_meta.length += worker_tensor_meta.length
+        ds_tensor_meta.max_shape = worker_tensor_meta.max_shape
+        ds_tensor_meta.min_shape = worker_tensor_meta.min_shape
+
+    # len of min_shape will be 0 if 0 outputs from worker
+    elif len(worker_tensor_meta.min_shape) != 0:
+        assert ds_tensor_meta.dtype == worker_tensor_meta.dtype
+        # TODO we can support this once we have ragged tensor support
+        assert len(ds_tensor_meta.max_shape) == len(worker_tensor_meta.max_shape)
+        assert len(ds_tensor_meta.min_shape) == len(worker_tensor_meta.min_shape)
+        ds_tensor_meta.length += worker_tensor_meta.length
+        ds_tensor_meta._update_shape_interval(tuple(worker_tensor_meta.max_shape))
+        ds_tensor_meta._update_shape_interval(tuple(worker_tensor_meta.min_shape))
+
+
+def merge_all_chunk_id_encoders(
+    all_workers_chunk_id_encoders: List[Dict[str, ChunkIdEncoder]],
+    ds_out: hub.core.dataset.Dataset,
+) -> None:
+    """Merges chunk_id_encoders from all workers into a single one and stores it in ds_out."""
+    tensors = list(ds_out.meta.tensors)
+    for tensor in tensors:
+        chunk_id_encoder = ds_out[tensor].chunk_engine.chunk_id_encoder
+        offset = chunk_id_encoder.num_samples
+
+        for current_worker_chunk_id_encoders in all_workers_chunk_id_encoders:
+            current_chunk_id_encoder = current_worker_chunk_id_encoders[tensor]
+            num_samples = current_chunk_id_encoder.num_samples
+            combine_chunk_id_encoders(
+                chunk_id_encoder, current_chunk_id_encoder, offset
+            )
+            offset += num_samples
+
+        chunk_id_key = get_chunk_id_encoder_key(tensor)
+        ds_out[tensor].chunk_engine.cache[chunk_id_key] = chunk_id_encoder
+    ds_out.flush()
+
+
+def combine_chunk_id_encoders(
+    ds_chunk_id_encoder: ChunkIdEncoder,
+    worker_chunk_id_encoder: ChunkIdEncoder,
+    offset: int,
+) -> None:
+    """Combines the dataset's chunk_id_encoder with a single worker's chunk_id_encoder."""
+    encoded_ids = worker_chunk_id_encoder._encoded
+    if encoded_ids.size != 0:
+        for encoded_id in encoded_ids:
+            encoded_id[1] += offset
+            if ds_chunk_id_encoder._encoded.size == 0:
+                ds_chunk_id_encoder._encoded = np.reshape(encoded_id, (-1, 2))
+            else:
+                ds_chunk_id_encoder._encoded = np.vstack(
+                    [ds_chunk_id_encoder._encoded, encoded_id]
+                )
```

Comment:
i feel like these methods should not exist in hub/util/transform, but rather the base encoder class. 

this seems pretty generalizable. 

for the most up to date base encoder, check out #1089. i don't recommend pulling in #1089 and sticking with main, but it's helpful to be aware of how im generalizing more complex stuff

Reply:
something like

```python
enc1 = SomeEncoderSubclass(...)
enc2 = SomeEncoderSubclass(...)
# ... encN ...

# ... add some encoding stuff ...

enc_main = enc1.merge(enc2, enc3, enc4, ...)  # ordering matters

Reply:
at the very least, try to generalize these functions and move them to another util file, like `hub/util/encoders` or something